10up/distributor
# Feature/858 Ability to set roles allowed to Pull

### Description of the Change

- This PR introduces a new setting to set user roles who are allowed to pull content.
- It also respects the `apply_filters( 'dt_pull_capabilities', 'manage_options' )` filter.
- If the current user has a cap coming from the filter, the current user will be allowed to pull despite the user roles settings.
- If the current user does not has a cap coming from the filter but has a user role that is allowed in the setting, then the current user can pull content.
- By default, `administrator` and `editor` roles are allowed to Pull content.
- The `dt_capabilities` filter is removed completelty.


Closes #858.

### Alternate Designs

![image](https://user-images.githubusercontent.com/25176325/165481600-fe29622f-09f8-4442-ba28-9fcd5d9963e8.png)
[image link](https://www.screencast.com/t/A7Db62H8NLI)

### Possible Drawbacks

- The `dt_capabilities` filter is removed completely, even if users stop using the filter, they will still see the main menu displayed. So far, the main `Distributor` menu was displayed via a filter (default to `manage_options`). But this PR removes this filter and displays the menu for all users by using the `read` cap.

https://github.com/10up/distributor/blob/08e6127e9a88c70560e59885171cf837540f2ce1/includes/external-connection-cpt.php#L663-L664

- Also removes the filter while displaying the submenu `Distributor > Pull Content` and displays it for all users by using the `read` cap. 

https://github.com/10up/distributor/blob/08e6127e9a88c70560e59885171cf837540f2ce1/includes/pull-ui.php#L154-L155

This was required because unlike the `Push Allowed` user role setting, we have to individually check each external connection to see if the current user can `Pull Content` or not. And to do so, we have to show the `Distributor > Pull Content` menu to all user roles.

### Verification Process

- Make sure you have at least 1 external connection set up.
- Make sure you have a few posts listed in `Distributor > Pull Content` and ready to be pulled. (If not, visit your external connection site, create a couple of posts and Distribute them to your original site, then you will see those posts are listed and ready to be pulled.)
- Please login as admin.
- Visit `Distributor > External Connections` and edit your external connection.
- See the `Roles Allowed to Pull` setting at the bottom.
- `Administrator` and `Editor` roles should be checked by default.
- Create an editor role user if does not exist already.
- Login to the editor role user in a separate browser (incognito recommended).
- See only the `Distributor > Pull Content` submenu is visible under the `Distributor`.
- Visit `Distributor > Pull Content`.
- See the posts listed - ready to be pulled.
- Try to pull 1 post, you should be able to pull successfully.
- Now go back to the browser/window having the admin role logged in.
- Visit `Distributor > External Connections` and edit your external connection.
- Untick the `Editor` role under the `Roles Allowed to Pull` setting at the bottom.
- Switch the browser/window having the editor role logged in.
- Visit `Distributor > Pull Content`.
- No posts should be listed!
- **Now let's check the filter:**
- Add the following code in your theme's `functions.php` file.
```
add_filter( 'dt_pull_capabilities', function() {
	return 'edit_others_posts';
} );
```
- Now even if the `Editor` role is unticked under the `Roles Allowed to Pull`, you should be able to see posts listed and ready to be pulled under the `Distributor > Pull Content`. This is because the editor role is now allowed to pull content via the filter.

### Checklist:






### Changelog Entry

> Added - Ability to set user roles to Pull Content
> Changed - The `Distributor > Pull Content` menu is now be visible for all user roles.
> Removed - The `dt_capabilities` & `dt_pull_capabilities` filters are removed while displaying the menus.

### Credits


Props @faisal-alvi @dkotter
